### PR TITLE
Adding GPU support

### PR DESCRIPTION
**Pull request by mclacore**

### CODE REVIEW

**Comment by mclacore** [comment]
Tested as working:
```
➜  azure-aks-cluster git:(michael/#ORC-436/gpu-support) ✗ k logs samples-tf-mnist-demo-tr4kx -f
2023-11-20 23:38:53.175993: I tensorflow/core/platform/cpu_feature_guard.cc:137] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX AVX2 FMA
2023-11-20 23:38:53.508860: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1030] Found device 0 with properties:
name: Tesla T4 major: 7 minor: 5 memoryClockRate(GHz): 1.59
pciBusID: 0001:00:00.0
totalMemory: 15.60GiB freeMemory: 15.50GiB
2023-11-20 23:38:53.508891: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1120] Creating TensorFlow device (/device:GPU:0) -> (device: 0, name: Tesla T4, pci bus id: 0001:00:00.0, compute capability: 7.5)
2023-11-20 23:42:47.991142: I tensorflow/stream_executor/dso_loader.cc:139] successfully opened CUDA library libcupti.so.8.0 locally
Successfully downloaded train-images-idx3-ubyte.gz 9912422 bytes.
Extracting /tmp/tensorflow/input_data/train-images-idx3-ubyte.gz
Successfully downloaded train-labels-idx1-ubyte.gz 28881 bytes.
Extracting /tmp/tensorflow/input_data/train-labels-idx1-ubyte.gz
Successfully downloaded t10k-images-idx3-ubyte.gz 1648877 bytes.
Extracting /tmp/tensorflow/input_data/t10k-images-idx3-ubyte.gz
Successfully downloaded t10k-labels-idx1-ubyte.gz 4542 bytes.
Extracting /tmp/tensorflow/input_data/t10k-labels-idx1-ubyte.gz
Accuracy at step 0: 0.0863
Accuracy at step 10: 0.713
Accuracy at step 20: 0.8263
Accuracy at step 30: 0.8641
Accuracy at step 40: 0.8784
Accuracy at step 50: 0.8896
Accuracy at step 60: 0.9028
Accuracy at step 70: 0.9108
Accuracy at step 80: 0.9102
Accuracy at step 90: 0.9156
Adding run metadata for 99
Accuracy at step 100: 0.9231
Accuracy at step 110: 0.9225
Accuracy at step 120: 0.927
Accuracy at step 130: 0.9294
Accuracy at step 140: 0.9296
Accuracy at step 150: 0.9308
Accuracy at step 160: 0.9335
Accuracy at step 170: 0.9322
Accuracy at step 180: 0.9346
Accuracy at step 190: 0.9371
Adding run metadata for 199
Accuracy at step 200: 0.9376
Accuracy at step 210: 0.9387
Accuracy at step 220: 0.9365
Accuracy at step 230: 0.9417
Accuracy at step 240: 0.943
Accuracy at step 250: 0.9409
Accuracy at step 260: 0.9424
Accuracy at step 270: 0.9413
Accuracy at step 280: 0.9429
Accuracy at step 290: 0.9464
Adding run metadata for 299
Accuracy at step 300: 0.9488
Accuracy at step 310: 0.944
Accuracy at step 320: 0.9466
Accuracy at step 330: 0.9494
Accuracy at step 340: 0.9495
Accuracy at step 350: 0.9455
Accuracy at step 360: 0.9443
Accuracy at step 370: 0.9475
Accuracy at step 380: 0.9461
Accuracy at step 390: 0.9478
Adding run metadata for 399
Accuracy at step 400: 0.951
Accuracy at step 410: 0.9536
Accuracy at step 420: 0.9564
Accuracy at step 430: 0.9543
Accuracy at step 440: 0.9545
Accuracy at step 450: 0.9554
Accuracy at step 460: 0.9536
Accuracy at step 470: 0.9575
```
Per: https://learn.microsoft.com/en-us/azure/aks/gpu-cluster#run-a-gpu-enabled-workload

File: src/main.tf [hunk 1]
```diff
@@ -43,13 +43,19 @@ resource "azurerm_kubernetes_cluster" "main" {
...
+    #bridgecrew:skip=BC_AZR_KUBERNETES_3:Skipping `Ensure AKS API server defines authorized IP ranges`
+    #bridgecrew:skip=BC_AZR_KUBERNETES_6:Skipping `Ensure AKS enables private clusters`
     name                        = var.node_groups.default_node_group.name
     vm_size                     = var.node_groups.default_node_group.node_size
     min_count                   = var.node_groups.default_node_group.min_size
     max_count                   = var.node_groups.default_node_group.max_size
     vnet_subnet_id              = var.vnet.data.infrastructure.default_subnet_id
     temporary_name_for_rotation = "${var.node_groups.default_node_group.name}temp"
     enable_auto_scaling         = true
+    max_pods                    = 50
```
Comment (chrisghill) [inline src/main.tf:58]: Why was this added? Have we ensured that the nodes can handle 50 pods? In EKS, the pod limit is based on the amount of ENI's that be attached, which changes based on the instance type.

File: core-services/nvidia_gpu.tf [hunk 2]
```diff
@@ -0,0 +1,80 @@
+resource "kubernetes_daemonset" "nvidia" {
+  count = try(var.node_groups.additional_node_groups.0.compute_type == "GPU" ? 1 : 0, 0)
+  metadata {
+    name      = "nvidia-device-plugin-daemonset"
+    namespace = kubernetes_namespace_v1.md-core-services.metadata.0.name
+    labels = {
```
Comment (chrisghill) [inline core-services/nvidia_gpu.tf:6]: We should be adding all of our default tags as labels

File: core-services/nvidia_gpu.tf [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+      match_labels = {
+        name = "nvidia-device-plugin-ds"
+      }
+    }
+    strategy {
+      type = "RollingUpdate"
+    }
+    template {
+      metadata {
+        labels = {
```
Comment (chrisghill) [inline core-services/nvidia_gpu.tf:21]: Same with pod labels. Without this, the pods won't be monitored.

File: massdriver.yaml [hunk 4]
```diff
@@ -32,49 +32,141 @@ NodeGroup: &node_group
       default: 10
       minimum: 1
       maximum: 1000
-    node_size:
+    compute_type:
```
Comment (chrisghill) [inline massdriver.yaml:35]: Is this change going to break existing AKS deployment configurations?

**Review by chrisghill — APPROVED** [review]

File: massdriver.yaml (hunk 4 above)
Comment (mclacore) [reply massdriver.yaml:35]: I tested this in my org and had no breaking issues. It shouldn't break because we're still using `node_size`, it just has different options based on the `compute_type`.

File: src/main.tf (hunk 1 above)
Comment (mclacore) [reply src/main.tf:58]: Added in via a bridgecrew suggestion. AKS can handle a lot more than 50: https://learn.microsoft.com/en-us/azure/aks/quotas-skus-regions#service-quotas-and-limits

**Review by chrisghill — APPROVED** [review]

File: massdriver.yaml (hunk 4 above)
Comment (chrisghill) [reply massdriver.yaml:35]: But the `node_size` schema is now dependent on the `compute_type`, which will be empty for all existing clusters. Not sure what this will do.

File: core-services/nvidia_gpu.tf (hunk 2 above)
Comment (mclacore) [reply core-services/nvidia_gpu.tf]: Added

File: core-services/nvidia_gpu.tf (hunk 3 above)
Comment (mclacore) [reply core-services/nvidia_gpu.tf]: Added

File: src/main.tf (hunk 1 above)
Comment (mclacore) [reply src/main.tf]: Removed this as it was a breaking change

File: massdriver.yaml (hunk 4 above)
Comment (mclacore) [reply massdriver.yaml:35]: Tested how this interaction would look for a customer. They would need to go into the config and select the `compute_type` and redeploy, but should not experience any downtime (even if selecting a different type/size)

computetype2
